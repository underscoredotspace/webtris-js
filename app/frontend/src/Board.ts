import { BoardGrid, Size } from "./types";

export default class Board {
  private grid: BoardGrid = [];
  private size: Size;

  constructor(w = 10, h = 19) {
    this.grid = new Array(h).fill(new Array(w).fill("x"));
    this.size = { w, h };
  }

  public setGrid(board: BoardGrid) {
    this.grid = board;
  }

  public copy(): BoardGrid {
    return this.grid.map((row) => row.slice());
  }

  public clearFullRows(): number {
    let lines = 0;
    let newBoardGrid: BoardGrid = [];
    for (let row of this.grid) {
      if (row.includes("x")) {
        newBoardGrid.push(row.slice());
      } else {
        newBoardGrid.unshift(new Array(this.size.w).fill("x"));
        lines++;
      }
    }
    this.grid = newBoardGrid;
    return lines;
  }
}
