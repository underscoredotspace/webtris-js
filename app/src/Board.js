import Row from '../src/Row'

export default class Board {
  constructor() {
    this.rows = []

    for (let ndx = 0; ndx<=17; ndx++) {
      this.rows[ndx] = new Row()
    }
  }

  render() {
    const rowHTML = this.rows.map(row => row.render()).join('')
    return `<div class="board">\n\t${rowHTML}\n</div>`
  }
}