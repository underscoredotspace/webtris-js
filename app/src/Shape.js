import shapes from './shapes';

export default class Shape {
  constructor() {
    const shape = shapes[Math.floor(Math.random() * shapes.length)]

    this.type = shape.type
    this.grid = shape.grid
    this.pos = {x: shape.start, y:2}
    this.rotateFix = shape.rotateFix
    this.rotateFixNdx = 0
  }

  move(x) {
    this.pos.x += x
  }

  drop() {
    this.pos.y++
  }

  unDrop() {
    this.pos.y--
  }

  rotateCW() {
    const grid = this.grid
    const newGrid = [...Array(grid[0].length)].map(e => Array(grid.length))

    for (let row in grid) {
      for (let col in grid[row]) {
        newGrid[col][grid.length-row-1] = grid[row][col]
      }
  }

    this.grid = newGrid

    this.rotateFixNdx++
    if (this.rotateFixNdx >= this.rotateFix.length) {this.rotateFixNdx = 0}

    this.pos.x += this.rotateFix[this.rotateFixNdx].x
    this.pos.y += this.rotateFix[this.rotateFixNdx].y
  }

  rotateCCW() {
    const grid = this.grid
    const newGrid = [...Array(grid[0].length)].map(e => Array(grid.length))

    for (let row in grid) {
      for (let col in grid[row]) {
        newGrid[grid[0].length-col-1][row] = grid[row][col]
      }
    }

    this.grid = newGrid

    this.pos.x -= this.rotateFix[this.rotateFixNdx].x
    this.pos.y -= this.rotateFix[this.rotateFixNdx].y

    this.rotateFixNdx--
    if (this.rotateFixNdx < 0) {this.rotateFixNdx = this.rotateFix.length-1}
  }
}