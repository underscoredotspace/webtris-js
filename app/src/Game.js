import Board from "./Board";
import Shape from "./Shape";

export default class Game {
  constructor() {
    this.board = new Board
    this.shape = new Shape
    this.nextShape = new Shape
    this.score = 0
    this.lines = 0
    this.atBottom = false
  }

  merged() {
    const board = this.board.copy()
    const shape = this.shape

    for (let row in shape.grid){
      for (let col in shape.grid[row]) {
        if (shape.grid[row][col] == 1) {
          board[Number(row)+shape.pos.y][Number(col)+shape.pos.x] = shape.type
        }
      }
    }

    return board
  }

  render() {
    const rows = this.merged()

    const boardHTML = ['<div class="board">']
    for (let row of rows) {
      if (this.atBottom && !row.includes('x')) {
        boardHTML.push('\t<div class="row full">\n')
      } else {
        boardHTML.push('\t<div class="row">\n')
      }
      for (let block of row) {
        boardHTML.push(`\t\t<div class="block" type="${block}"></div>\n`)
      }
      boardHTML.push('\t</div>\n')
    }

    boardHTML.push('</div>')

    return boardHTML.join('')
} 
}