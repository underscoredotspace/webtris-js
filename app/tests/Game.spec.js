import Game from '../src/Game'
import Board from '../src/Board';
import Shape from '../src/Shape';

describe("Game()", () => {
  let game

  beforeEach(() => {
    game = new Game()
  })

  test("Starting point",  () => {
    expect(game.board).toBeInstanceOf(Board)
    expect(game.shape).toBeInstanceOf(Shape)
    expect(game.nextShape).toBeInstanceOf(Shape)
    expect(game.score).toBe(0)
    expect(game.lines).toBe(0)
  })

  test("merged()", () => {
    const mergedBoard = game.merged()
    expect(mergedBoard).toMatchSnapshot()
  })

  test("render() with no full rows", () => {
    expect(game.render()).toMatchSnapshot()
  })

  test("render() with a full row", () => {
    game.board.grid[18] = new Array(10).fill('t')
    game.atBottom = true
    expect(game.render()).toMatchSnapshot()
  })
})