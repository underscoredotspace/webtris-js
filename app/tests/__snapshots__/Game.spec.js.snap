// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game() merged() 1`] = `
Array [
  Array [
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
  ],
  Array [
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
  ],
  Array [
    "x",
    "x",
    "x",
    "t",
    "t",
    "t",
    "x",
    "x",
    "x",
    "x",
  ],
  Array [
    "x",
    "x",
    "x",
    "x",
    "t",
    "x",
    "x",
    "x",
    "x",
    "x",
  ],
  Array [
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
  ],
  Array [
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
  ],
  Array [
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
  ],
  Array [
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
  ],
  Array [
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
  ],
  Array [
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
  ],
  Array [
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
  ],
  Array [
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
  ],
  Array [
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
  ],
  Array [
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
  ],
  Array [
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
  ],
  Array [
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
  ],
  Array [
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
  ],
  Array [
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
  ],
  Array [
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
    "x",
  ],
]
`;

exports[`Game() render() with a full row 1`] = `
"<div class=\\"board\\">	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"t\\"></div>
		<div class=\\"block\\" type=\\"t\\"></div>
		<div class=\\"block\\" type=\\"t\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"t\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row full\\">
		<div class=\\"block\\" type=\\"t\\"></div>
		<div class=\\"block\\" type=\\"t\\"></div>
		<div class=\\"block\\" type=\\"t\\"></div>
		<div class=\\"block\\" type=\\"t\\"></div>
		<div class=\\"block\\" type=\\"t\\"></div>
		<div class=\\"block\\" type=\\"t\\"></div>
		<div class=\\"block\\" type=\\"t\\"></div>
		<div class=\\"block\\" type=\\"t\\"></div>
		<div class=\\"block\\" type=\\"t\\"></div>
		<div class=\\"block\\" type=\\"t\\"></div>
	</div>
</div>"
`;

exports[`Game() render() with no full rows 1`] = `
"<div class=\\"board\\">	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"t\\"></div>
		<div class=\\"block\\" type=\\"t\\"></div>
		<div class=\\"block\\" type=\\"t\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"t\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
	<div class=\\"row\\">
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
		<div class=\\"block\\" type=\\"x\\"></div>
	</div>
</div>"
`;
